### PR TITLE
[Enhancement] limit buffer capacity instead of scan concurrency

### PR DESCRIPTION
**Pull request by ZiheLiu**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Problem Summary(Required) ：

Use the dynamic statistics of chunk memory usage to limit the capacity of a chunk buffer instead of scan concurrency.
1. Before triggering a new I/O task, pin a chunk token to pin a position in the buffer.
2. Before creating a new chunk in the I/O task, pin a chunk token, except the first new chunk (because it already pined a token in the step 1).
3. After a chunk is popped from buffer, unpin the token.

### CODE REVIEW

**Comment by ZiheLiu** [comment]
run starrocks_be_unittest

File: be/src/exec/vectorized/olap_scan_node.cpp [hunk 1]
```diff
@@ -714,7 +715,7 @@ void OlapScanNode::_estimate_scan_and_output_row_bytes() {
 }
 
 size_t OlapScanNode::_scanner_concurrency() const {
```
Comment (mofeiatwork) [inline be/src/exec/vectorized/olap_scan_node.cpp:717]: Is this method still used?

File: be/src/exec/pipeline/scan/chunk_buffer_limiter.h [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+    ChunkBufferTokenPtr pin(int num_chunks) override;
+
+    bool is_full() const override { return _acquired_chunks_counter >= _capacity; }
+    size_t size() const override { return _acquired_chunks_counter; }
+    size_t capacity() const override { return _capacity; }
+    size_t default_capacity() const override { return _default_capacity; }
+
+private:
+    void _unpin(int num_chunks) { _acquired_chunks_counter.fetch_sub(num_chunks); }
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/scan/chunk_buffer_limiter.h:109]: how about add some `DCHECK` here to make sure the correctness of `pin` and `unpin` mechanism ?

**Review by murphyatwork — DISMISSED** [review]

File: be/src/exec/vectorized/olap_scan_node.cpp (hunk 1 above)
Comment (ZiheLiu) [reply be/src/exec/vectorized/olap_scan_node.cpp:717]: Yes, non-pipeline uses this.

**Review by murphyatwork — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**